### PR TITLE
gnome: Fix glib tool lookup consistency

### PR DESCRIPTION
**Pull request by xclaesse**
It was not always using the pkg-config file.

### CODE REVIEW

**Comment by dabrain34** [comment]
File "C:\Users\scerveau\Documents\DEV\meson\mesonbuild\interpreterbase\decorators.py", line 581, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
  File "C:\Users\scerveau\Documents\DEV\meson\mesonbuild\modules\gnome.py", line 1880, in mkenums_simple
    c_file = self._make_mkenum_impl(state, kwargs['sources'], body_filename, c_cmd)
  File "C:\Users\scerveau\Documents\DEV\meson\mesonbuild\modules\gnome.py", line 1932, in _make_mkenum_impl
    real_cmd: T.List[T.Union[str, ExternalProgram]] = [self._find_tool(state, 'glib-mkenums')]
NameError: name 'self' is not defined

subprojects\gstreamer\gst\meson.build:212:0: ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.

**Comment by xclaesse** [comment]
@dabrain34 yeah, saw that in the CI too... fixed now.

**Comment by dabrain34** [comment]
```
  File "C:\Users\scerveau\Documents\DEV\meson\mesonbuild\modules\gnome.py", line 1882, in mkenums_simple
    c_file = self._make_mkenum_impl(state, kwargs['sources'], body_filename, c_cmd)
  File "C:\Users\scerveau\Documents\DEV\meson\mesonbuild\modules\gnome.py", line 1934, in _make_mkenum_impl
    real_cmd: T.List[T.Union[str, 'ToolType']] = [self._find_tool(state, 'glib-mkenums')]
  File "C:\Users\scerveau\Documents\DEV\meson\mesonbuild\modules\gnome.py", line 344, in _find_tool
    pcfile = tool_map[tool]
KeyError: 'glib-mkenums'

subprojects\gstreamer\gst\meson.build:212:0: ERROR: Unhandled python exception
```
is it expected ?

**Comment by xclaesse** [comment]
@dabrain34 argh, a typo, fixed now.

**Comment by tristan957** [comment]
Looks like this just needs a review?

**Comment by dabrain34** [comment]
Unfortunately the bug still happens on my windows setup where pkconfig finds the pc file for glib but the binaries are not in PATH

**Comment by tristan957** [comment]
It would be good if you could grab a print out of the path Meson is trying to search for the tools in

**Comment by xclaesse** [comment]
> Looks like this just needs a review?

yes please

> Unfortunately the bug still happens on my windows setup where pkconfig finds the pc file for glib but the binaries are not in PATH

Would need the new logs to check what's going on.

**Comment by tristan957** [comment]
Maybe this should be marked for 0.64.1 if the fix is confirmed

**Comment by dabrain34** [comment]
Here it is:

[meson-log.txt](https://github.com/mesonbuild/meson/files/9955507/meson-log.txt)
[configure_gnome_tool.txt](https://github.com/mesonbuild/meson/files/9955505/configure_gnome_tool.txt)

**Comment by nirbheek** [comment]
> Howerver, later gst-plugins-base lookup glib-mkenums manually with `glib_mkenums = find_program('glib-mkenums')` and that's not using the one from pkg-config.

I think the fix there is probably just to port that code to use `gnome.mkenums_simple()`? No need to add code to meson for something that is a build file mistake.

**Comment by tristan957** [comment]
@nirbheek @xclaesse what is left on this PR?

**Comment by xclaesse** [comment]
Rebased, I think it was ready to be merged but then got forgotten.

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -203,6 +203,8 @@ class MkEnums(_MkEnumsCommon):
         vtail: T.Optional[str]
         depends: T.List[T.Union[BuildTarget, CustomTarget, CustomTargetIndex]]
 
+    ToolType = T.Union['Executable', ExternalProgram, 'OverrideProgram']
```
Comment (tristan957) [inline mesonbuild/modules/gnome.py:206]: I would remove the quotes around types. Seems inconsistent in various places now.

File: mesonbuild/modules/gnome.py [hunk 2]
```diff
@@ -317,6 +319,21 @@ def _print_gdbus_warning() -> None:
                      mlog.bold('https://github.com/mesonbuild/meson/issues/1387'),
                      once=True, fatal=False)
 
+    def _find_tool(self, state: 'ModuleState', tool: str) -> 'ToolType':
```
Comment (tristan957) [inline mesonbuild/modules/gnome.py:322]: Make this a static method

File: mesonbuild/modules/gnome.py [hunk 3]
```diff
@@ -317,6 +319,21 @@ def _print_gdbus_warning() -> None:
...
+            'gio-querymodules': 'gio-2.0',
+            'glib-compile-schemas': 'gio-2.0',
+            'glib-compile-resources': 'gio-2.0',
+            'gdbus-codegen': 'gio-2.0',
+            'glib-genmarshal': 'glib-2.0',
+            'glib-mkenums': 'glib-2.0',
+            'g-ir-scanner': 'gobject-introspection-1.0',
+            'g-ir-compiler': 'gobject-introspection-1.0',
+        }
+        pcfile = tool_map[tool]
```
Comment (tristan957) [inline mesonbuild/modules/gnome.py:333]: I believe they are called modules in pkgconfig, right? Not files.

File: mesonbuild/modules/gnome.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/modules/gnome.py]: done

File: mesonbuild/modules/gnome.py (hunk 2 above)
Comment (xclaesse) [reply mesonbuild/modules/gnome.py]: done

**Review by tristan957 — APPROVED** [review]

File: mesonbuild/modules/gnome.py (hunk 3 above)
Comment (xclaesse) [reply mesonbuild/modules/gnome.py]: Renamed to depname which is what's used for state.find_tool() argument as well.

**Comment by fanc999-1** [comment]
Hi,

Hmm, seems that this update here caused a problem like this when building GNOME-using items using `cmd.exe` consoles:

```
ninja: error: 'c:/vs14.0/x64/bin/glib-compile-resources', needed by 'gdk/gdkresources.h', missing and no known rule to make it
```
(notice that the `.exe` for `glib-compile-resources` is not there, and Ninja expects it to be so)

Is it preferable that we update GLib when generating the pkg-config file, or append `.exe` here ourselves, if needed, for maximum compatibility with older GLib releases (or even, the `glib-genmarshal` which was formerly a C-based executable).

With blessings, thank you!

**Comment by eli-schwartz** [comment]
@fanc999-1 it is fixed on git master with #12316 and this PR is queued to be backported.

**Comment by fanc999-1** [comment]
Hi @eli-schwartz,

Thanks a bunch!

**Pull request merged**